docs/cookbook/evals/evaluating_web_search_agent.md

```diff
@@ -157,11 +161,64 @@ class WebAssistant(BaseModel):
             print()
 ```
 
-For more information regarding this code snippet, refer to the Web Search Agent cookbook.
+## Basic Evaluations
+
+Let's start off with some basic evaluations to know whether our agent is working in general. Given that updates to prompts can significantly influence LLM behavior, it's crucial to test each component of our agent individually.
+
+### Evaluating `_web_search` tool
+
+Our goal is to ensure that the LLM consistently utilizes the web search tool, rather than relying on its inherent knowledge base to generate responses. We've intentionally refrained from explicitly instructing the agent to always utilize the web search tool, as some user queries may be more conversational in nature and not necessitate web searches. However, for user queries that are more information-seeking, the agent should always leverage the web search tool.
+
+```python
+import pytest
+
+user_queries = [
+    "How is the weather in New York?",
+    "What is the capital of France?",
+    "Who is the president of the United States?",
+    "What is the population of India?",
+    "What is an apple?",
+]
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize("user_query", user_queries)
+async def test_web_search(user_query: str):
+    """Tests that the web search agent always uses the web search tool."""
+    web_assistant = WebAssistant()
+    response = await web_assistant._stream(user_query)
+    async for _, tool in response:
+        assert tool._name() == "_web_search" if tool else False
```

Comment:
hmm, maybe we should instead do something like this?

```python
web_assistant = WebAssistant()
response = await web_assistant._stream(user_query)
tools = []
async for _, tool in response:
    if tool:
        tools.append(tool)
assert len(tools) == 1  and tools[0]._name() == "_web_search"
```

Reply:
added, but is there any harm with the llm calling the web search tool multiple times?

Reply:
hmm, it's not that there's harm there so much as what the expected behavior is? I would be surprised if it needs to call `_web_search` multiple times since the tool accepts a list of queries. I would generally expect it to call the tool, then parse the webpages, and then make a new web search call if the initial parsing doesn't pan out, right?

If the `_web_search` tool only took a single query as input, then maybe it would make multiple calls and we could just assert `len(tools) > 1`

Reply:
okay it does call once majority of times